specs/phase0/beacon-chain.md

```diff
@@ -1589,16 +1584,45 @@ def process_final_updates(state: BeaconState) -> None:
             or validator.effective_balance + UPWARD_THRESHOLD < balance
         ):
             validator.effective_balance = min(balance - balance % EFFECTIVE_BALANCE_INCREMENT, MAX_EFFECTIVE_BALANCE)
-    # Reset slashings
+```
+
+```python
+def process_slashings_updates(state: BeaconState) -> None:
+    next_epoch = Epoch(get_current_epoch(state) + 1)
     state.slashings[next_epoch % EPOCHS_PER_SLASHINGS_VECTOR] = Gwei(0)
-    # Set randao mix
+```
+
+```python
+def process_randao_mixes_updates(state: BeaconState) -> None:
+    current_epoch = get_current_epoch(state)
+    next_epoch = Epoch(current_epoch + 1)
     state.randao_mixes[next_epoch % EPOCHS_PER_HISTORICAL_VECTOR] = get_randao_mix(state, current_epoch)
-    # Set historical root accumulator
+```
+
+```python
+def process_historical_roots_updates(state: BeaconState) -> None:
+    next_epoch = Epoch(get_current_epoch(state) + 1)
     if next_epoch % (SLOTS_PER_HISTORICAL_ROOT // SLOTS_PER_EPOCH) == 0:
         historical_batch = HistoricalBatch(block_roots=state.block_roots, state_roots=state.state_roots)
         state.historical_roots.append(hash_tree_root(historical_batch))
-    # Rotate participation records
-    rotate_participation_records(state)
+```
+
+
+```python
+def process_participation_record_updates(state: BeaconState) -> None:
+    # Rotate current/previous epoch attestations
+    state.previous_epoch_attestations = state.current_epoch_attestations
+    state.current_epoch_attestations = []
+```
+
+```python
+def process_final_updates(state: BeaconState) -> None:
```

Comment:
IMPORTANT: This should be entirely non-substantive. Any substantive change is a _mistake_.

Needed more granular control on ordering of these components in phase 0 vs phase 1. Things like `process_participation_record_updates` can't happen before `process_online_tracking` because the rewards flags need to still be in place for both epochs